src/sinks/opendal_common.rs

```diff
@@ -0,0 +1,292 @@
+    type Events = Vec<Event>;
+    type Encoder = (Transformer, Encoder<Framer>);
+    type Payload = Bytes;
+    type Request = OpenDALRequest;
+    type Error = std::io::Error;
+
+    fn compression(&self) -> Compression {
+        self.compression
+    }
+
+    fn encoder(&self) -> &Self::Encoder {
+        &self.encoder
+    }
+
+    fn split_input(
+        &self,
+        input: (String, Vec<Event>),
+    ) -> (Self::Metadata, RequestMetadataBuilder, Self::Events) {
+        let (partition_key, mut events) = input;
+        let finalizers = events.take_finalizers();
+        let opendal_metadata = OpenDALMetadata {
+            partition_key,
+            count: events.len(),
+            byte_size: events.size_of(),
+            finalizers,
+        };
+
+        let builder = RequestMetadataBuilder::from_events(&events);
+
+        (opendal_metadata, builder, events)
+    }
+
+    fn build_request(
+        &self,
+        mut metadata: Self::Metadata,
+        request_metadata: RequestMetadata,
+        payload: EncodeResult<Self::Payload>,
+    ) -> Self::Request {
+        // TODO: we can support time format later.
+        let name = uuid::Uuid::new_v4().to_string();
+        let extension = self.compression.extension();
+
+        metadata.partition_key = format!("{}{}.{}", metadata.partition_key, name, extension);
+
+        OpenDALRequest {
+            metadata,
+            payload: payload.into_payload(),
+            request_metadata: request_metadata,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct OpenDALResponse {
+    pub count: usize,
+    pub events_byte_size: usize,
+    pub byte_size: usize,
+}
+
+impl DriverResponse for OpenDALResponse {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(self.count, self.events_byte_size)
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.byte_size)
+    }
+}
+
+impl Service<OpenDALRequest> for OpenDALService {
+    type Response = OpenDALResponse;
+    type Error = opendal::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    // Emission of an internal event in case of errors is handled upstream by the caller.
+    fn poll_ready(&mut self, _: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    // Emission of internal events for errors and dropped events is handled upstream by the caller.
+    fn call(&mut self, request: OpenDALRequest) -> Self::Future {
+        let byte_size = request.payload.len();
+        let op = self.op.clone();
+
+        Box::pin(async move {
+            let result = op
+                .object(&request.metadata.partition_key.as_str())
+                .write(request.payload)
+                .in_current_span()
+                .await;
+            result.map(|_| {
+                emit!(OpenDALBytesSent { byte_size });
```

Comment:
This isn't necessary as the Driver will emit a `BytesSent` for us due to the `bytes_sent` implementation in `OpenDALResponse`. You will also need to set the protocol on the driver at line 105.

[see here](https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/stream/driver.rs#L219-L221)

```suggestion
```

Reply:
Thanks a lot!